### PR TITLE
Dont fail when datasource or dashboard already exists

### PR DESCRIPTION
**Pull request by caruccio**

### CODE REVIEW

**Comment by caruccio** [comment]
Hey @vrutkovs do you know why is it failing tests? Looks like to me it's something unrelated to the commit itself.

**Comment by vrutkovs** [comment]
@caruccio yeah, these failures are flaking Origin tests, not related to this PR. Lets try some more

/retest

if that doesn't work I'll merge it manually after a few attempts

**Comment by caruccio** [comment]
Hey @mjudeikis is it possible to merge it?

**Comment by mjudeikis** [comment]
ping @sdodson

**Comment by vrutkovs** [comment]
/hold

@mrsiano still has comments about rewriting the dashboard

**Comment by caruccio** [comment]
Oops, my bad. Didn't saw that comment.

**Comment by sdodson** [comment]
> /cherrypick release-3.9

The whole grafana stack is tech preview. I'd prefer we do not backport changes to previous GA releases, I'm fine with doing it for current GA but it's too much of a burden to go any further back for a component that's not officially supported.

File: roles/openshift_grafana/tasks/install_grafana.yaml [hunk 1]
```diff
@@ -175,6 +175,9 @@
     method: POST
     body: '{{ payload_data }}'
     body_format: json
+    status_code:
+    - 200
+    - 409
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply roles/openshift_grafana/tasks/install_grafana.yaml]: >Dashboards tend to have very personal tweak

User should not be modifying ansible-managed dashboard and expect the changes will persist after ansible rerun.

**Comment by caruccio** [comment]
@vrutkovs @sdodson would you guys mind taking another look? Both datasource and dashboards are overwritten.

File: roles/openshift_grafana/tasks/install_grafana.yaml [hunk 2]
```diff
@@ -213,14 +242,20 @@
     src: "{{ cluster_monitoring_dashboard }}"
   register: slurpfile
 
+- set_fact:
+    dashboard_data: '{{ slurpfile["content"] | b64decode | from_json | combine({ "dashboard": { "overwrite": true } }, recursive=True) }}'
```
[the remark this answers is not in the document]
Comment (caruccio) [reply roles/openshift_grafana/tasks/install_grafana.yaml]: Done

File: roles/openshift_grafana/tasks/install_grafana.yaml [hunk 3]
```diff
@@ -230,14 +265,20 @@
     src: "{{ node_exporter_dashboard }}"
   register: slurpfile
 
+- set_fact:
```
[the remark this answers is not in the document]
Comment (caruccio) [reply roles/openshift_grafana/tasks/install_grafana.yaml]: Done

File: roles/openshift_grafana/tasks/install_grafana.yaml [hunk 4]
```diff
@@ -175,11 +175,40 @@
...
+      validate_certs: false
+      method: PUT
+      body: '{{ payload_data }}'
+      body_format: json
+      headers:
+        Content-Type: "Content-Type: application/json"
+      status_code:
+      - 200
+    register: update_ds
+  - debug: var=update_ds
```
[the remark this answers is not in the document]
Comment (caruccio) [reply roles/openshift_grafana/tasks/install_grafana.yaml]: Done

File: roles/openshift_grafana/tasks/install_grafana.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrsiano) [reply roles/openshift_grafana/tasks/install_grafana.yaml]: @vrutkovs @caruccio i think we would like to know about relevant metrics that are missing, and if so contribute them to the project.
this project should give the end user the relevant graphs he needed.

AFAIK in the openshift-monitoring we'll not allow to modify the graphs

**Comment by vrutkovs** [comment]
Works fine here, two more things left to do:
* [ ] Overwrite dashboards we create
  Similar to datasource - if the dashboard is created by ansible it should be managed by ansible. User might want to restore the default by rerunning it. A cloned dashboard should be used to persist changes.
* [ ] Squash commits

**Comment by mrsiano** [comment]
if we're not going to overwrite the dashboard than ltgm otherwise update the pr to overwrite the dashboard

**Comment by vrutkovs** [comment]
Okay, lets not overwrite the dashboard then - it would be easier to contribute back to the original dashboard (tweaks, fixes etc.). If the user would require resetting the dashboard it should be removed - and ansible rerun would fix that.

Setting
/approve
here, however lets squash commits and I'll set LGTM to get things moving further

**Comment by caruccio** [comment]
In fact, the commit e43bb0485aaadd4a90d825aec5938140e7f573e9 already overwrites both dashboards (via `dashboard.overwtite=true` post data) and datasource (does by default, when accepting status 409).

Should I remove dashboard overwrite or keep it prior to squashing?

**Comment by mrsiano** [comment]
I think it should be fine..
/lgtm

**Pull request merged**